src/libcrun/kontain_nonkmexec.c

```diff
@@ -0,0 +1,347 @@
+ * Entries in the file are a single line with fields separated by colons.
+ * There are 3 fields:
+ *   regular expression to match an input string against
+ *   the file to exec to if the re matches
+ *   a sha of the file being exec'ed to to verify the file is unchanged since the
+ *   db was created.
+ * Failures is this function cause the process to exit.
+ *
+ * We should probably use some sort of json representation for this instead of the
+ * archaic colon separated fields.
+ */
+static void
+kontain_execpath_builddb(execpath_state_t* statep)
+{
+   char buf[1024];
+   FILE* f = NULL;
+   char *fields[3];
+   execpath_entry_t* epp = NULL;
+   execpath_entry_t* tail = NULL;
+   int rc = 0;
+
+   if (statep->execpath_dbpath != NULL) {
+      kontain_execpath_freedb(statep);
+   }
+   f = fopen(statep->execpath_dbpath, "r");
+   if (f == NULL) {
+      libcrun_fail_with_error(errno, "Couldn't open %s", statep->execpath_dbpath);
+   }
+   buf[sizeof(buf)-1] = 0;
+   while (fgets(buf, sizeof(buf), f) != NULL) {
+      if (buf[sizeof(buf)-1] != 0) {
+         // a really long line.
+         rc = E2BIG;
+         buf[sizeof(buf)-1] = 0;
+         libcrun_fail_with_error(rc, "execpath db line too long: %s", buf);
+      }
+      if (buf[0] == '#') {  // ignore comments
+         continue;
+      }
+      char* s = strchr(buf, '\n');
+      if (s != NULL) {
+         *s = 0;
+      }
+      // Split into fields
+      if (split_into_fields(buf, fields) != 0) {
+         rc = EINVAL;
+         libcrun_fail_with_error(rc, "Couldn't split <%s> into fields", buf);
+      }
+
+      epp = calloc(sizeof(execpath_entry_t), 1);
+      if (epp == NULL) {
+         libcrun_fail_with_error(ENOMEM, "Couldn't allocate %d bytes of memory", sizeof(execpath_entry_t));
+      }
+
+      // Compile regular expression.
+      rc = regcomp(&epp->re, fields[0], REG_EXTENDED | REG_NOSUB);
+      if (rc != 0) {
+         char regerrbuf[128];
+         regerror(rc, &epp->re, regerrbuf, sizeof(regerrbuf));
+         libcrun_fail_with_error(rc, "Couldn't compile regular expression %s, %s", fields[0], regerrbuf);
+      }
+      epp->path = strdup(fields[1]);
+      if (epp->path == NULL) {
+         rc = ENOMEM;
+         libcrun_fail_with_error(rc, "Couldn't strdup %s", fields[1]);
+      }
+
+      // Compute hash of target file
+      rc = sha256_file(fields[1], epp->sha256);
+      if (rc != 0) {
+         libcrun_fail_with_error(rc, "Couldn't compute hash of file %s", fields[1]);
+      }
+      // Does the hash match what the exec_ok file has?
+      if (strcmp(epp->sha256, fields[2]) != 0) {
+         printf("%s: Computed hash %s doesn't match expected hash %s\n",
+                fields[1],
+                epp->sha256,
+                fields[2]);
+         libcrun_fail_with_error(EILSEQ, "hash on file %s differ: expected: %s got: %s", epp->path, fields[2], epp->sha256);
+      }
+
+      // Chain entry on to the end of the list
```

Comment:
is there a way in crun  to trace this ? Just a log record ? would be good to have explicit records of paths added to the list

Reply:
There is an error and warning logging interface that does not abort the command.

So, you would prefer a warning and return indicating the executable must be run with km.

Reply:
I'd prefer a trace (it's not a warning, it's normal trace)

- if the config is wrong, warn, ignore and continue
- if the config is correct, trace and continue

Reply:
crun doesn't seem to have a normal information trace.
The libcrun_warning() function is the one that seems to do what is needed but the message can be in the color red or labled as a warning if using syslog or journald.

Reply:
ok then